boa/src/syntax/parser/statement/declaration/hoistable/generator_decl/mod.rs

```diff
@@ -0,0 +1,122 @@
+#[cfg(test)]
+mod tests;
+
+use crate::syntax::{
+    ast::{node::declaration::generator_decl::GeneratorDecl, Keyword, Punctuator},
+    parser::{
+        function::FormalParameters,
+        function::FunctionBody,
+        statement::{BindingIdentifier, LexError, Position},
+        AllowAwait, AllowDefault, AllowYield, Cursor, ParseError, TokenParser,
+    },
+};
+use std::io::Read;
+
+/// Generator declaration parsing.
+///
+/// More information:
+///  - [MDN documentation][mdn]
+///  - [ECMAScript specification][spec]
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*
+/// [spec]: https://tc39.es/ecma262/#prod-GeneratorDeclaration
+#[derive(Debug, Clone, Copy)]
+pub(super) struct GeneratorDeclaration {
+    allow_yield: AllowYield,
+    allow_await: AllowAwait,
+    is_default: AllowDefault,
+}
+
+impl GeneratorDeclaration {
+    /// Creates a new `GeneratorDeclaration` parser.
+    pub(super) fn new<Y, A, D>(allow_yield: Y, allow_await: A, is_default: D) -> Self
+    where
+        Y: Into<AllowYield>,
+        A: Into<AllowAwait>,
+        D: Into<AllowDefault>,
+    {
+        Self {
+            allow_yield: allow_yield.into(),
+            allow_await: allow_await.into(),
+            is_default: is_default.into(),
+        }
+    }
+}
+
+impl<R> TokenParser<R> for GeneratorDeclaration
+where
+    R: Read,
+{
+    type Output = GeneratorDecl;
+
+    fn parse(self, cursor: &mut Cursor<R>) -> Result<Self::Output, ParseError> {
```

Comment:
Wouldn't it be possible to extract a common function for `GeneratorDeclaration` and `FunctionDeclaration`? The code is practically the same, and we can return `name`, `params` and `body` from the function as a tuple.

Reply:
We could do that, but there would be some annoying `if`s around that make everything less readable. We would also have to pass in a flag to signal if it can be only a function/generator like [here](https://tc39.es/ecma262/#prod-LabelledItem). Also the error messages would be a bit messy. I would say the better readability is worth the few copied lines? What do you think?

Reply:
I think it's worth it, because in `AsyncFunctionDeclaration` we already have `// TODO: If self.is_default, then this can be empty.` solved, but because the three parsers are disjointed we didn't realize that we should copy this logic on the other parsers.

Also, we don't need to make a common function for the whole `parse`, we can get the biggest subset of instructions that are the same between the three, possibly passing `function`, `async function` and `generator` as `&str` to show good error messages.